<a href="https://colab.research.google.com/github/vishwanathbeena/appliedai_practice/blob/main/Bootstrap_assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# <font color='red'><b>Bootstrap assignment</b> </font>

<b>There will be some functions that start with the word "grader" ex: grader_sampples(), grader_30().. etc, you should not change those function definition.

Every Grader function has to return True.</b>

<font color='blue'> <b>Importing packages</b> </font>

In [ ]:
import numpy as np # importing numpy for numerical computation
from sklearn.datasets import load_boston # here we are using sklearn's boston dataset
from sklearn.metrics import mean_squared_error # importing mean_squared_error metric

In [ ]:
boston = load_boston()
x=boston.data #independent variables
y=boston.target #target variable

In [ ]:
print(x.shape)
print(y.shape)
print(x[0])
print(y[:5])

(506, 13)
(506,)
[6.320e-03 1.800e+01 2.310e+00 0.000e+00 5.380e-01 6.575e+00 6.520e+01
 4.090e+00 1.000e+00 2.960e+02 1.530e+01 3.969e+02 4.980e+00]
[24.  21.6 34.7 33.4 36.2]


In [ ]:
x[:5]

array([[6.3200e-03, 1.8000e+01, 2.3100e+00, 0.0000e+00, 5.3800e-01,
        6.5750e+00, 6.5200e+01, 4.0900e+00, 1.0000e+00, 2.9600e+02,
        1.5300e+01, 3.9690e+02, 4.9800e+00],
       [2.7310e-02, 0.0000e+00, 7.0700e+00, 0.0000e+00, 4.6900e-01,
        6.4210e+00, 7.8900e+01, 4.9671e+00, 2.0000e+00, 2.4200e+02,
        1.7800e+01, 3.9690e+02, 9.1400e+00],
       [2.7290e-02, 0.0000e+00, 7.0700e+00, 0.0000e+00, 4.6900e-01,
        7.1850e+00, 6.1100e+01, 4.9671e+00, 2.0000e+00, 2.4200e+02,
        1.7800e+01, 3.9283e+02, 4.0300e+00],
       [3.2370e-02, 0.0000e+00, 2.1800e+00, 0.0000e+00, 4.5800e-01,
        6.9980e+00, 4.5800e+01, 6.0622e+00, 3.0000e+00, 2.2200e+02,
        1.8700e+01, 3.9463e+02, 2.9400e+00],
       [6.9050e-02, 0.0000e+00, 2.1800e+00, 0.0000e+00, 4.5800e-01,
        7.1470e+00, 5.4200e+01, 6.0622e+00, 3.0000e+00, 2.2200e+02,
        1.8700e+01, 3.9690e+02, 5.3300e+00]])

## <font color='red'><b>Task 1</b></font>

<font color='red'> <b>Step - 1</b></font>

*  <font color='blue'><b>Creating samples</b></font><br>
    <b> Randomly create 30 samples from the whole boston data points</b>
    *  Creating each sample: Consider any random 303(60% of 506) data points from whole data set and then replicate any 203 points from the sampled points
    
     For better understanding of this procedure lets check this examples, assume we have 10 data points [1,2,3,4,5,6,7,8,9,10], first we take 6 data points randomly , consider we have selected [4, 5, 7, 8, 9, 3] now we will replicate 4 points from [4, 5, 7, 8, 9, 3], consder they are [5, 8, 3,7] so our final sample will be [4, 5, 7, 8, 9, 3, 5, 8, 3,7]
* <font color='blue'><b> Create 30 samples </b></font>
    *  Note that as a part of the Bagging when you are taking the random samples <b>make sure each of the sample will have different set of columns</b><br>
Ex: Assume we have 10 columns[1 ,2 ,3 ,4 ,5 ,6 ,7 ,8 ,9 ,10] for the first sample we will select [3, 4, 5, 9, 1, 2] and for the second sample  [7, 9, 1, 4, 5, 6, 2] and so on...
Make sure each sample will have atleast 3 feautres/columns/attributes

<font color='red'><b>Step - 2 </b></font>

<font color='blue'><b>Building High Variance Models on each of the sample and finding train MSE value</b></font>

*  Build a regression trees on each of 30 samples.
*  Computed the predicted values of each data point(506 data points) in your corpus.
*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{30}\sum_{k=1}^{30}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$
*  Now calculate the $MSE =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$

<font color='red'> <b>Step - 3 </b></font>

*  <font color='blue'><b>Calculating the OOB score </b></font>

*  Predicted house price of $i^{th}$ data point $y^{i}_{pred} =  \frac{1}{k}\sum_{\text{k= model which was buit on samples not included } x^{i}}(\text{predicted value of } x^{i} \text{ with } k^{th} \text{ model})$.
*  Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

# <font color='red'><b>Task 2</b></font>

*  <font color='blue'><b>Computing CI of OOB Score and Train MSE</b></font>
  *   Repeat Task 1 for 35 times, and for each iteration store the Train MSE and OOB score </li>
<li> After this we will have 35 Train MSE values and 35 OOB scores </li>
<li> using these 35 values (assume like a sample) find the confidence intravels of MSE and OOB Score </li>
<li> you need to report CI of MSE and CI of OOB Score </li>
<li> Note: Refer the Central_Limit_theorem.ipynb to check how to find the confidence intravel</li>
</ol>

# <font color='red'><b>Task 3</b></font>

*  <font color='blue'><b>Given a single query point predict the price of house.</b></font>

Consider xq= [0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60] 
Predict the house price for this point as mentioned in the step 2 of Task 1.

<br><br><br>

# <font color='blue'> <b>Task - 1</b></font>

<font color='blue'><b>Step - 1</b></font>

*  <font color='blue'> <b>Creating samples</b></font>

<font color='Orange'><b>Algorithm</b></font>

![alt text](https://i.imgur.com/BTVYXQ1.jpg/)

*  <font color='blue'><b> Write code for generating samples</b></font>

In [ ]:
from scipy import sparse
def generating_samples(input_data, target_data):

    '''In this function, we will write code for generating 30 samples '''
    # you can use random.choice to generate random indices without replacement
    # Please have a look at this link https://docs.scipy.org/doc/numpy-1.16.0/reference/generated/numpy.random.choice.html for more details
    # Please follow above pseudo code for generating samples 
    

    # return sampled_input_data , sampled_target_data,selected_rows,selected_columns
    #note please return as lists

    random_row_index_303 = np.random.choice(range(506), 303, replace=False)
    # random_row_index_303 = np.random.randint(0,507,size=303)
    rand_col_size = np.random.randint(3,13,size=1)
    random_col_index = np.random.choice(range(13),size=rand_col_size,replace=False)

    selected_rows = input_data[np.ix_(random_row_index_303,random_col_index)]
    target_303 = target_data[random_row_index_303]

    random_index_203 = np.random.randint(0,203,size=203)
    replacing_rows = selected_rows[random_index_203,:] 
    target_203 = target_data[random_index_203]
    final_sample_data = np.vstack((selected_rows,replacing_rows))
    final_target_data = np.vstack((target_303.reshape(-1,1),target_203.reshape(-1,1)))
    return final_sample_data.tolist(),final_target_data.tolist(),random_row_index_303.tolist(),random_col_index.tolist()


In [ ]:
a,b,c,d = generating_samples(x,y)
print(len(a))

506


<font color='cyan'> <b> Grader function - 1 </b> </fongt>

In [ ]:
def grader_samples(a,b,c,d):
    length = (len(a)==506  and len(b)==506)
    sampled = (len(a)-len(set([str(i) for i in a]))==203)
    rows_length = (len(c)==303)
    column_length= (len(d)>=3)
    assert(length and sampled and rows_length and column_length)
    return True
a,b,c,d = generating_samples(x, y)
grader_samples(a,b,c,d)

True

*  <font color='blue'> <b>Create 30 samples </b>

![alt text](https://i.imgur.com/p8eZaWL.jpg)

In [ ]:
# Use generating_samples function to create 30 samples 
# store these created samples in a list
def get_sample_lists(x,y):
  list_input_data =[]
  list_output_data =[]
  list_selected_row= []
  list_selected_columns=[]
  for i in range(30):
    a,b,c,d = generating_samples(x, y)
    list_input_data.append(a)
    list_output_data.append(b)
    list_selected_row.append(c)
    list_selected_columns.append(d)
  return list_input_data,list_output_data,list_selected_row,list_selected_columns

In [ ]:
list_input_data,list_output_data,list_selected_row,list_selected_columns = get_sample_lists(x,y)

<font color='cyan'> <b>Grader function - 2 </b></font>

In [ ]:
def grader_30(a):
    assert(len(a)==30 and len(a[0])==506)
    return True
grader_30(list_input_data)

True

<br>

<font color='red'><b>Step - 2 </b></font>

<font color='orange'><b>Flowchart for building tree</b></font>

![alt text](https://i.imgur.com/pcXfSmp.png)

*  <font color='blue'><b> Write code for building regression trees</b></font>

In [ ]:
from sklearn.tree import DecisionTreeRegressor
def get_model_lists(list_input_data,list_output_data):
  model_list = []
  for i in range(len(list_input_data)):
    clf = DecisionTreeRegressor()
    clf.fit(list_input_data[i],list_output_data[i])
    model_list.append(clf)
  return model_list



<font color='orange'><b>Flowchart for calculating MSE </b></font>

![alt text](https://i.imgur.com/sPEE618.png)

After getting predicted_y for each data point, we can use sklearns mean_squared_error to calculate the MSE between predicted_y and actual_y.

*  <font color='blue'><b> Write code for calculating MSE</b></font>

In [ ]:
def get_MSE(model_list,list_selected_columns,x,y):
  prediction_list=[]
  for i in range(len(model_list)):
    predict = model_list[i].predict(x[:,list_selected_columns[i]])
    prediction_list.append(predict.tolist())
  predict_array_np = np.array(prediction_list)
  final_predict_np = (np.sum(predict_array_np,axis=0))/(len(prediction_list))
  MSE = np.sum(np.power((y - final_predict_np ),2))/(len(final_predict_np))
  return MSE

<font color='blue'><b>Step - 3 </b></font>

<font color='orange'><b>Flowchart for calculating OOB score</b></font>

![alt text](https://i.imgur.com/95S5Mtm.png)

Now calculate the $OOB Score =  \frac{1}{506}\sum_{i=1}^{506}(y^{i} - y^{i}_{pred})^{2}$.

*  <font color='blue'><b> Write code for calculating OOB score </b></font>

In [ ]:
def get_oob_score(x,y,list_selected_columns,model_list,list_selected_row):
  data_point_pred = []
  for i in range(len(x)):
    counter = 0
    model_pred = 0
    for j in range(len(model_list)):
      if i not in list_selected_row[j]:
        counter+=1
        model_pred += model_list[j].predict(x[np.ix_([i],list_selected_columns[j])])
    data_point_pred.append((model_pred/counter)[0])
  OOB = np.sum(np.power((y - data_point_pred ),2))/(len(data_point_pred))
  return OOB

In [ ]:
def get_mse_oob(x,y):
  list_input_data,list_output_data,list_selected_row,list_selected_columns = get_sample_lists(x,y)
  model_list = get_model_lists(list_input_data,list_output_data)
  mse = get_MSE(model_list,list_selected_columns,x,y)
  oob = get_oob_score(x,y,list_selected_columns,model_list,list_selected_row)
  return mse,oob


In [ ]:
mse,oob = get_mse_oob(x,y)
print('MSE is {}'.format(mse))
print('OOB is {}'.format(oob))

MSE is 10.043598392914559
OOB is 21.493751688540556


# <font color='blue'><b>Task 2</b></font>

In [ ]:
mse_list= []
oob_list= []
for i in range(35):
  mse,oob = get_mse_oob(x,y)
  mse_list.append(mse)
  oob_list.append(oob)



In [ ]:
print(mse_list)
print(oob_list)
print(len(oob_list))
print(len(mse_list))

[12.170201575959851, 10.04047784332869, 10.529150790393125, 10.083957038485295, 11.303042307134701, 10.351742651419919, 9.071782089210402, 11.200779386850051, 10.69608230904442, 9.932303854841848, 11.826218094248091, 10.697667336533978, 10.473931885273606, 10.407463331557746, 10.990160110190326, 10.53093588025954, 10.604856957908046, 11.39683267853905, 11.95902572165145, 9.587461324310093, 9.932846320090444, 9.711132669848068, 9.585771013462638, 10.785655604741901, 9.706481934025696, 10.624458752384214, 10.672374207775775, 10.111817049483765, 12.36151400028749, 10.53259776890965, 9.75912169646993, 9.709466372405537, 10.211851803387319, 11.207196309493549, 11.007261701932526]
[26.467546106637773, 21.70274842745948, 21.981377343414927, 21.970609485824795, 23.354294867386653, 22.99999148866988, 20.198106765547067, 23.136441385550647, 24.540644158624424, 21.91311122787284, 24.934025314513026, 22.812925443380482, 24.473429964826842, 22.929174475729877, 23.811333295872455, 22.463275908761453

In [ ]:
mse_mean = np.mean(mse_list)
mse_std_dev = np.std(mse_list)
oob_std_dev = np.std(oob_list)
oob_mean = np.mean(oob_list)
print(mse_mean)
print(mse_std_dev)
print(oob_std_dev)
print(oob_mean)
MSE_conf_interval_ub = mse_mean+2*(mse_std_dev/np.sqrt(35))
MSE_conf_interval_lb = mse_mean-2*(mse_std_dev/np.sqrt(35))
oob_conf_interval_ub = oob_mean+2*(oob_std_dev/np.sqrt(35))
oob_conf_interval_lb = oob_mean-2*(oob_std_dev/np.sqrt(35))
print(MSE_conf_interval_lb,MSE_conf_interval_ub)
print(oob_conf_interval_lb,oob_conf_interval_ub)


10.564960582052537
0.7667779028752442
1.5201777613090146
22.829495590133885
10.305742339233808 10.824178824871266
22.315581708968292 23.343409471299477


# <font color='blue'><b>Task 3</b></font>

<font color='orange'><b>Flowchart for Task 3</b></font>

<b>Hint: </b> We created 30 models by using 30 samples in TASK-1. Here, we need send query point "xq"  to 30 models and perform the regression on the output generated by 30 models.

![alt text](https://i.imgur.com/Y5cNhQk.png)

*  <font color='blue'><b> Write code for TASK 3 </b></font>

In [ ]:
list_input_data,list_output_data,list_selected_row,list_selected_columns = get_sample_lists(x,y)
model_list = get_model_lists(list_input_data,list_output_data)
def get_prediction_for_data_point(x):
  prediction_list= []
  for i in range(len(model_list)):
    predict = model_list[i].predict(x[:,list_selected_columns[i]])
    prediction_list.append(predict.tolist())
  predict_array = np.array(prediction_list)
  final_predict = (np.sum(predict_array,axis=0))/(len(prediction_list))
  return final_predict

data_prediction = get_prediction_for_data_point(np.array([[0.18,20.0,5.00,0.0,0.421,5.60,72.2,7.95,7.0,30.0,19.1,372.13,18.60]]))
print(data_prediction)


[22.43888889]


In [ ]:
print(np.mean(y))
print(np.std(y))

22.532806324110677
9.188011545278203


<br><br><br>

<font color='red'><b>Write observations for task 1, task 2, task 3 indetail</b></font>

Observations:
1. When trained using 30 samples and tested MSE is MSE is 10.043598392914559 and OOB is 21.493751688540556

2. 95 % confidence interval of 35 different samples of MSE is [10.305742339233808 10.824178824871266] and OOB is [22.315581708968292 23.343409471299477].

BothMSE and OOB of task-1 are not in range of 95% CI obtained in TASk-2.This could be for reasons that samples taken in task-1 ar just 30.But where as in task-2 we did the experiment 35 times by taking sample sizes of 30.according to centallimit theorem, this should reflect population MSE and OOB more accurately.

The  prediction obtained for singel data point is 22.43, which is around the range of mean of target variable.

